src/lib/yang/xpath.lua

```diff
@@ -0,0 +1,283 @@
+-- and "routes.route". If the end is reached it'll return nil.
+local function next_element(path)
+   return string.match(path, "([^/]+)/?(.*)")
+end
+
+local function split_path(path)
+   local tail = path
+   return function ()
+      local head
+      head, tail = next_element(tail)
+      if head == nil then return head else return extract_parts(head) end
+   end
+end
+
+-- Finds the grammar node for a fragment in a given grammar.
+local function extract_grammar_node(grammar, fragment)
+   local errmsg = "Invalid path: "..fragment.name
+   if grammar.type == "table" then
+      if grammar.keys[fragment.name] == nil then
+         return assert(grammar.values[fragment.name], errmsg)
+      else
+         return grammar.keys[fragment.name]
+      end
+   else
+      return assert(grammar[fragment.name], errmsg)
+   end
+end
+
+-- Converts an XPath path to a lua array consisting of path componants.
+-- A path component can then be resolved on a yang data tree:
+function convert_path(grammar, path)
+   local ret = {}
+   local node = grammar
+   for element in split_path(path) do
+      node = extract_grammar_node(node, element)
+      local luapath, next_node = handle(node.type, element, node)
+      table.insert(ret, luapath)
+      node = next_node
+   end
+   return ret
+end
+
+-- Returns a resolver for a paticular schema and *lua* path.
+function resolve(schema, path)
+   local schema = lib.deepcopy(schema)
+   local path = lib.deepcopy(path)
+   local grammar = datalib.data_grammar_from_schema(schema)
+   local handlers = {}
+   local function handle(scm, prod, data, path)
+      if #path == 0 then return data end
+      return assert(handlers[scm.kind], scm.kind)(scm, prod, data, path)
+   end
+   function handlers.list(scm, prod, data, path)
+      local head = table.remove(path, 1)
+      -- The list can either be a ctable or a plain old lua table, if it's the
+      -- ctable then it requires some more work to retrive the data.
+      local prod = prod.members[head.name]
+      local data = data[normalize_id(head.name)]
+      if #prod.keys > 1 then error("Multiple key values are not supported!") end
+      if prod.key_ctype then
+         -- It's a ctable and we need to prepare the key so we can lookup the
+         -- pointer to the entry and then convert the values back to lua.
+         local kparser = valuelib.types[scm.body[scm.key].primitive_type].parse
+         local key = kparser(head.keys[scm.key])
+         local ckey = ffi.new(prod.key_ctype)
+         ckey[scm.key] = key
+
+         data = data:lookup_ptr(ckey).value
+      else
+         data = data[normalize_id(head.keys[scm.key])]
+      end
+
+      if #path == 0 then return data end
+      local peek = path[1]
+      if type(peek) == "table" then peek = peek.name end
+      scm = scm.body[peek]
+      prod = prod.values[peek]
+      return handle(scm, prod, data, path)
+   end
+   function handlers.container(scm, prod, data, path)
```

Comment:
So for the handlers in the resolving code I think we need to operate on the schema to be able to find the primitive type in the list handler for example. The other handles otherwise don't really use it and just advance it on.

I can't think of a way to re-write the list handler without the schema but i'm open to suggestions if I've missed something?